README.md

```diff
@@ -1,2 +1,202 @@
+# `@consento/promise`
+
+A set of utilities for working with [Promises]() with particular focus
+on abort-ability using [`AbortSignal`s][AbortSignal].
+
+[Promise]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
+[AbortSignal]: https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal
+
+## Goal
+
+Provide sturdy `Promise` utils that make it easier to write asynchronous code that is both performant
+and saves on unnecessary operations by making it easier to use AbortSignals.
+
+The library has a partial focus on `Typescript` utility as `TypeScript` comes with its own
+set of problems.
+
+## Contributions
+
+Additionally to the common Consento contribution guidelines, this repository is eager to be dissolved
+into standard projects. If you find one of the utils presented here useful in a more common context (e.g.
+Node.js repo, TypeScript or w3c) then we would find it awesome if you could promote it to find its way there so
+we can remove it from this library.
+
+We also appreciate pointers to projects that may offer a better solution than the solutions presented
+in this repo!
+
+## Utils
+
+- [`bubbleAbort()`](#bubble-abort) - tiny helper for dealing with abort signals
+- [`checkpoint()`](#checkpoint) - comfortable helper for dealing with abort signals
+- [`cbPromise()`](#cb-promise) - tiny helper for asynchronous code
+- [`cleanupPromise()`](#cleanup-promise) - versatily promise variant that allows react-style hook cleanups
+- [`composeAbort()`](#compose-abort) - compose an abort-signal into sub-variants.
+- [`isPromiseLike()`](#is-promise-like) - typescript type-check helper
+- [`raceWithSignal()`](#race-with-signal) - races several promises, all but the first will be aborted
+- [`toPromise()`](#to-promise) - turns a value into a promise, cleverly
+- [`wrapTimeout()`](#wrap-timeout) - wraps an async function to have an optional timeout
+
+#### `bubbleAbort()`
+
+Simple function that throws an AbortError if an AbortSignal
+happens to be aborted.
+
+_Usage:_
+
+```javascript
+const { bubbleAbort } = require('@consento/promise/bubbleAbort')
+
+async function longRunning ({ signal } = {}) {
+  let result = ''
+  for await (const data of internalIterator()) {
+    bubbleAbort(signal) // Will throw an error if the signal is aborted
+    result += data
+  }
+  return result
+}
+```
+
+#### `cbPromise()`
+
+Extended Promise that has the `resolve` and `reject` methods
+on the class instance instead of a template function.
+
+```javascript
+const { CBPromise } = require('@consento/promise/cbPromise')
+
+const p = new CBPromise()
+p.cb(new Error(), 'result') // traditional callback API
+p.resolve('result') // or use traditional resolve
+p.reject(new Error()) // and reject
+```
+
+This may come in handy when you need a callback but want to await
+the promise.
+
+```javascript
+async function test () {
+  const p = new CBPromise()
+  setTimeout(p.resolve, 500)
+  await p
+}
+```
+
+#### `checkpoint()`
```

Comment:
Not clear what the arguments are here. Does it take some data and return that same data?

Reply:
Exactly, it takes the data awaits and returns it but will throw an error if aborted.